File: app/src/main/java/com/yawar/memo/views/UserInformationActivity.java
package com.yawar.memo.views;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.yawar.memo.call.RequestCallActivity;
import com.yawar.memo.modelView.UserInformationViewModel;
import com.yawar.memo.service.SocketIOService;
import com.yawar.memo.sessionManager.ClassSharedPreferences;
import com.yawar.memo.Api.ServerApi;
import com.yawar.memo.R;
import com.yawar.memo.adapter.MediaAdapter;
import com.yawar.memo.constant.AllConstants;
import com.yawar.memo.model.MediaModel;
import com.yawar.memo.model.UserModel;
import com.yawar.memo.repositry.BlockUserRepo;
import com.yawar.memo.repositry.ChatRoomRepo;
import com.yawar.memo.utils.BaseApp;
import com.yawar.memo.utils.TimeProperties;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import de.hdodenhof.circleimageview.CircleImageView;

public class UserInformationActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private ArrayList<MediaModel> recyclerDataArrayList = new ArrayList<>();
    ServerApi serverApi;
    CircleImageView circleImageView;
    LinearLayout linerMore, linerAudioCall,linerVideoCall;
    TextView txtUserName;
    TextView txtSpecialNumber;
    TextView txtState;
    TimeProperties timeProperties;

    BlockUserRepo blockUserRepo;
    UserInformationViewModel userInformationViewModel;
    TextView call;
    TextView video;
    TextView message;
    TextView mute;
    TextView more;
    TextView special_number;
    TextView media;
    String userName;
    String sn;
    String chatId;
    String another_user_id;
    String my_id;
    String fcm_token;
    String imageUrl;
    ClassSharedPreferences classSharedPreferences;
    BaseApp myBase;
    MediaAdapter adapter;
    ImageView imgBtnMessage;
    PopupMenu p;
    ChatRoomRepo chatRoomRepo;
    String blockedFor;
    boolean isBlockForMe = false;
    ImageView imageVideoCall, imageAudioCall, imageMore, imageChat;
    public static final String CHEK = "ConversationActivity.CHECK_CONNECT";


//    float textSize = 14.0F ;
//    SharedPreferences sharedPreferences ;


    private final BroadcastReceiver check = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String check = intent.getExtras().getString("check");
            JSONObject checkObject = null;
            String checkConnect = "false";
            String user_id;

            try {

                checkObject = new JSONObject(check);
                user_id = checkObject.getString("user_id");
                if(user_id.equals(another_user_id)) {
                    checkConnect = checkObject.getString("is_connect");
                    userInformationViewModel.setLastSeen(checkObject.getString("last_seen"));
                    userInformationViewModel.set_state(checkConnect);
                }
                } catch (JSONException e) {
                e.printStackTrace();
            }


        }
    };
    private void sendBlockFor(Boolean blocked) {

        JSONObject userBlocked = new JSONObject();
        JSONObject item = new JSONObject();

        try {
            item.put("blocked_for",userInformationViewModel.blockedFor().getValue());
            item.put("Block",blocked);
            userBlocked.put("my_id", my_id);
            userBlocked.put("user_id",another_user_id );
            userBlocked.put("blocked_for",userInformationViewModel.blockedFor().getValue());
            userBlocked.put("userDoBlockName",classSharedPreferences.getUser().getUserName());
            userBlocked.put("userDoBlockSpecialNumber",classSharedPreferences.getUser().getSecretNumber());
            userBlocked.put("userDoBlockImage",classSharedPreferences.getUser().getImage());


        } catch (JSONException e) {
            e.printStackTrace();
        }
        Intent service = new Intent(this, SocketIOService.class);
        service.putExtra(SocketIOService.EXTRA_BLOCK_PARAMTERS, userBlocked.toString());
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_BLOCK);
        startService(service);


    }

    private void sendUnBlockFor(Boolean blocked) {

        JSONObject userUnBlocked = new JSONObject();

        try {
            userUnBlocked.put("my_id", my_id);
            userUnBlocked.put("user_id",another_user_id );
            userUnBlocked.put("blocked_for",userInformationViewModel.blockedFor().getValue());

        } catch (JSONException e) {
            e.printStackTrace();
        }
        Intent service = new Intent(this, SocketIOService.class);


        service.putExtra(SocketIOService.EXTRA_UN_BLOCK_PARAMTERS, userUnBlocked.toString());
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_UN_BLOCK);
        startService(service);


    }
    private void checkConnect() {
        Intent service = new Intent(this, SocketIOService.class);
        JSONObject object = new JSONObject();
        try {
            object.put("my_id", my_id);
            object.put("your_id", another_user_id);
//            socket.emit("check connect", object);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        service.putExtra(SocketIOService.EXTRA_CHECK_CONNECT_PARAMTERS, object.toString());
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_CHECK_CONNECT);
        startService(service);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_user_information);
        LocalBroadcastManager.getInstance(this).registerReceiver(check, new IntentFilter(CHEK));
        recyclerView = findViewById(R.id.idCourseRV);
         adapter = new MediaAdapter(recyclerDataArrayList, this);


        // setting grid layout manager to implement grid view.
        // in this method '2' represents number of columns to be displayed in grid view.
        LinearLayoutManager layoutManager = new LinearLayoutManager(this,  LinearLayoutManager.HORIZONTAL,false);

        // at last set adapter to recycler view.
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);
        initViews();
        initAction();
    }



    private void initViews() {
        classSharedPreferences = BaseApp.getInstance().getClassSharedPreferences();
        userInformationViewModel = new ViewModelProvider(this).get(UserInformationViewModel.class);
        timeProperties = new TimeProperties();


        myBase = BaseApp.getInstance();
        chatRoomRepo = myBase.getChatRoomRepo();
        blockUserRepo= myBase.getBlockUserRepo();
        Bundle bundle = getIntent().getExtras();
       userName = bundle.getString("name", "Default");
       sn = bundle.getString("special", "Default");
       chatId = bundle.getString("chat_id", "Default");
       another_user_id = bundle.getString("user_id", "Default");
        fcm_token = bundle.getString("fcm_token", "Default");
        imageUrl = bundle.getString("image", "Default");
        blockedFor = bundle.getString("blockedFor","");
        my_id = classSharedPreferences.getUser().getUserId();
        checkConnect();
        userInformationViewModel.setBlockedFor(blockedFor);
        txtState = findViewById(R.id.last_seen);
        userInformationViewModel.state.observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                System.out.println("stateee"+s);
                if(s!=null){
                    if (s.equals("true")) {
//                        isCoonect = true;
                        System.out.println("dialay"+s);
                        txtState.setText(R.string.connect_now);
                    } else if (s.equals("false")) {
//                        isCoonect = false;

                        if (!userInformationViewModel.getLastSeen().equals("null")) {
                            txtState.setText(getResources().getString(R.string.last_seen) + " " + timeProperties.getDateForLastSeen(UserInformationActivity.this, Long.parseLong(userInformationViewModel.getLastSeen())));
                        }


                    }

                }
            }
        });

        userInformationViewModel.getMedia().observe(this, new androidx.lifecycle.Observer<ArrayList<MediaModel>>() {
                    @Override
                    public void onChanged(ArrayList<MediaModel> mediaModelArrayList) {
                        if (mediaModelArrayList != null) {

                            for (MediaModel media:
                                 mediaModelArrayList) {
                                recyclerDataArrayList.add(media);
                            }


                        }
                        adapter.notifyDataSetChanged();
                    }
                });



//////////////////////////
        userInformationViewModel.isBlocked().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean s) {
                if(s!=null){
                    sendBlockFor(s);
                    userInformationViewModel.setBlocked(null);


                }
            }
        });
        ////////////
        userInformationViewModel.isUnBlocked().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean s) {
                System.out.println("stateee"+s);
                if(s!=null){
//                    conversationModelView.
                    sendUnBlockFor(s);
                    userInformationViewModel.setUnBlocked(null);


                }
            }
        });
        userInformationViewModel.blockedFor().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                boolean isAnyOneBlock = false;
                if (s != null) {
                    if (s.equals(my_id)||s.equals("0")) {
                        isBlockForMe = true;
                        imageAudioCall.setEnabled(false);
                        imageVideoCall.setEnabled(false);
           }
                    else if (s.equals(another_user_id)) {

                        isBlockForMe = false;
                        imageAudioCall.setEnabled(false);
                        imageVideoCall.setEnabled(false);


                    }
             else {
                        isBlockForMe = false;
                        imageAudioCall.setEnabled(true);
                        imageVideoCall.setEnabled(true);


                    }
            }
                else{
                    isBlockForMe = false;
                    imageAudioCall.setEnabled(true);
                    imageVideoCall.setEnabled(true);

                }

        }});

        /////////



        serverApi = new ServerApi(this);
        linerMore = findViewById(R.id.liner_more);
        linerVideoCall=findViewById(R.id.liner_video_call);
        linerAudioCall=findViewById(R.id.liner_audio_call);
        imageVideoCall=findViewById(R.id.img_video_call);
        imageAudioCall=findViewById(R.id.img_audio_call);
        imageMore=findViewById(R.id.img_more);
        imageChat=findViewById(R.id.img_message);












        circleImageView = findViewById(R.id.imageView);
        txtUserName = findViewById(R.id.txt_user_name);
//        txtUserName.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

        txtSpecialNumber = findViewById(R.id.txt_special_number);
//        txtSpecialNumber.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

        call = findViewById(R.id.call);
//        call.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

        video = findViewById(R.id.video);
//        video.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

        message = findViewById(R.id.message);
//        message.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

//        mute = findViewById(R.id.mute);
//        mute.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));


        more = findViewById(R.id.more);
//        more.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

        special_number = findViewById(R.id.special_number);
//        special_number.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

        media = findViewById(R.id.media);
//        media.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));



        if(!imageUrl.isEmpty()){
            Glide.with(circleImageView.getContext()).load(AllConstants.imageUrl+imageUrl).into(circleImageView);}
       txtUserName.setText(userName);
        System.out.println(sn+"special_number");
        if(!sn.isEmpty()) {
            String firstString = sn.substring(0, 1);
            String secondString = sn.substring(1, 4);
            String thirtyString = sn.substring(4, 7);
            String lastString = sn.substring(7);

            txtSpecialNumber.setText(firstString + "-" + secondString + "-" + thirtyString + "-" + lastString);
        }
        userInformationViewModel.mediaRequest(my_id,another_user_id);
//        getMedia();




    }
    private void initAction() {
        imageChat.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Bundle bundle = new Bundle();


                bundle.putString("reciver_id",another_user_id);

                bundle.putString("sender_id", my_id);
                bundle.putString("fcm_token",fcm_token );

//        bundle.putString("reciver_id",chatRoomModel.reciverId);
                bundle.putString("name",userName);
                bundle.putString("image",imageUrl);
                bundle.putString("chat_id",chatId);
                bundle.putString("blockedFor", userInformationViewModel.blockedFor().getValue());



                Intent intent = new Intent(UserInformationActivity.this, ConversationActivity.class);
                intent.putExtras(bundle);

                startActivity(intent);
            }
        });
        //////
        imageMore.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
//                block();
                popupMenuExample();

            }
        });
        imageAudioCall.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
//                block();
                Intent intent = new Intent(UserInformationActivity.this, RequestCallActivity.class);
                intent.putExtra("anthor_user_id", another_user_id);
                intent.putExtra("user_name", userName);
                intent.putExtra("isVideo", false);
                intent.putExtra("fcm_token", fcm_token);
                intent.putExtra("image_profile", imageUrl);



                startActivity(intent);
            }
        });

        imageVideoCall.setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(View view) {
//                block();
            Intent intent = new Intent(UserInformationActivity.this, RequestCallActivity.class);
//                Intent intent = new Intent(ConversationActivity.this, CompleteActivity.class);
            intent.putExtra("anthor_user_id", another_user_id);
            intent.putExtra("user_name", userName);
            intent.putExtra("isVideo", true);
            intent.putExtra("fcm_token", fcm_token);
            intent.putExtra("image_profile", imageUrl);
            startActivity(intent);

        }
    });
}


    private void popupMenuExample() {
        PopupMenu p = new PopupMenu(this, linerMore);
        p.getMenuInflater().inflate(R.menu.main_menu, p .getMenu());

        if(isBlockForMe){
            p.getMenu().findItem(R.id.block).setVisible(false);
            p.getMenu().findItem(R.id.unBlock).setVisible(true);}
        else {
            p.getMenu().findItem(R.id.block).setVisible(true);
            p.getMenu().findItem(R.id.unBlock).setVisible(false);
        }

        p.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            public boolean onMenuItemClick(MenuItem item) {
                UserModel userModel = new UserModel(another_user_id,userName,userName,"","", sn,imageUrl,"");

                switch (item.getItemId()){
                    case R.id.block:
                        AlertDialog.Builder dialog=new AlertDialog.Builder(UserInformationActivity.this);
                        dialog.setTitle(R.string.alert_block_user);
                        dialog.setPositiveButton(R.string.block,
                                new DialogInterface.OnClickListener() {
                                    public void onClick(DialogInterface dialog,
                                                        int which) {

//                                        serverApi.block(my_id,userModel);
                                        userInformationViewModel.sendBlockRequest(my_id,another_user_id);
                                    }
                                });
                        dialog.setNegativeButton(R.string.cancel,new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                            }
                        });

                        AlertDialog alertDialog=dialog.create();
                        alertDialog.show();

                        break;

//
                    case R.id.unBlock:

                    AlertDialog.Builder dialogUnBlock=new AlertDialog.Builder(UserInformationActivity.this);
                        dialogUnBlock.setTitle(R.string.alert_unblock_user);
                        dialogUnBlock.setPositiveButton(R.string.Unblock,
                            new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog,
                                                    int which) {

//                                    serverApi.unbBlockUser(my_id,userModel);
                                    userInformationViewModel.sendUnBlockRequest(my_id,another_user_id);

                                }
                            });
                        dialogUnBlock.setNegativeButton(R.string.cancel,new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();
                        }
                    });

                    AlertDialog alertUnBlockDialog=dialogUnBlock.create();
                        alertUnBlockDialog.show();

                    break;


//                        p.getMenu().findItem(R.id.block).setVisible(false);
//                        p.getMenu().findItem(R.id.block).setVisible(true);




                }
                return true;
            }
        });
        p.show();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(check);

    }
}

File: app/src/main/java/com/yawar/memo/ui/userInformationPage/UserInformationActivity.kt
package com.yawar.memo.ui.userInformationPage

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.app.Dialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.PopupMenu
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.rxjava2.RxDataStore
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.github.chrisbanes.photoview.PhotoView
import com.yawar.memo.R
import com.yawar.memo.ui.userInformationPage.media.MediaAdapter
import com.yawar.memo.utils.CallProperty
import com.yawar.memo.ui.requestCall.RequestCallActivity
import com.yawar.memo.constant.AllConstants
import com.yawar.memo.databinding.ActivityUserInformationBinding
import com.yawar.memo.model.MediaModel
import com.yawar.memo.model.UserModel
import com.yawar.memo.repositry.BlockUserRepo
import com.yawar.memo.service.SocketIOService
import com.yawar.memo.sessionManager.ClassSharedPreferences
import com.yawar.memo.ui.chatPage.ConversationActivity
import com.yawar.memo.ui.chatPage.ConversationModelView
import com.yawar.memo.ui.introPage.IntroActModelView
import com.yawar.memo.utils.BaseApp
import com.yawar.memo.utils.TimeProperties
import com.yawar.memo.utils.checkThereIsOngoingCall
import dagger.hilt.android.AndroidEntryPoint
import org.json.JSONException
import org.json.JSONObject
import javax.inject.Inject

@AndroidEntryPoint
class UserInformationActivity : AppCompatActivity() {
//    var EXAMPLE_COUNTER: Key<Int>
    private val recyclerDataArrayList = ArrayList<MediaModel>()
//    lateinit  var userInformationViewModelFactory: UserInformationViewModelFactory
    lateinit var timeProperties: TimeProperties
    lateinit var dataStore: RxDataStore<Preferences>
    lateinit var binding : ActivityUserInformationBinding
//    lateinit var sharedPreferenceLiveData: SharedPreferenceStringLiveData
    var muteList: ArrayList<String?>? = ArrayList()
    @Inject
    lateinit var blockUserRepo: BlockUserRepo
     val userInformationViewModel by viewModels<UserInformationViewModel>()
    lateinit var userName: String
    lateinit var sn: String
    lateinit var chatId: String
    lateinit var another_user_id: String
    lateinit var my_id: String
    lateinit var fcm_token: String
    lateinit var imageUrl: String
    lateinit var classSharedPreferences: ClassSharedPreferences
    lateinit var myBase: BaseApp
    var adapter: MediaAdapter? = null
    var p: PopupMenu? = null

    //    ChatRoomRepo chatRoomRepo;
    lateinit var blockedFor: String
    var isBlockForMe = false

    private val check: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val check = intent.extras!!.getString("check")
            var checkObject: JSONObject? = null
            var checkConnect: String? = "false"
            val user_id: String
            try {
                checkObject = JSONObject(check)
                user_id = checkObject.getString("user_id")
                if (user_id == another_user_id) {
                    checkConnect = checkObject.getString("is_connect")
                    userInformationViewModel.setLastSeen(checkObject.getString("last_seen"))
                    userInformationViewModel.set_state(checkConnect)
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    private fun sendBlockFor(blocked: Boolean) {
        val userBlocked = JSONObject()
        val item = JSONObject()
        try {
            item.put("blocked_for", userInformationViewModel.blockedFor().value)
            item.put("Block", blocked)
            userBlocked.put("my_id", my_id)
            userBlocked.put("user_id", another_user_id)
            userBlocked.put("blocked_for", userInformationViewModel.blockedFor().value)
            userBlocked.put("userDoBlockName", classSharedPreferences.user.userName)
            userBlocked.put("userDoBlockSpecialNumber", classSharedPreferences.user.secretNumber)
            userBlocked.put("userDoBlockImage", classSharedPreferences.user.image)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        val service = Intent(this, SocketIOService::class.java)
        service.putExtra(SocketIOService.EXTRA_BLOCK_PARAMTERS, userBlocked.toString())
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_BLOCK)
        startService(service)
    }

    private fun sendUnBlockFor(blocked: Boolean) {
        val userUnBlocked = JSONObject()
        try {
            userUnBlocked.put("my_id", my_id)
            userUnBlocked.put("user_id", another_user_id)
            userUnBlocked.put("blocked_for", userInformationViewModel.blockedFor().value)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        val service = Intent(this, SocketIOService::class.java)
        service.putExtra(SocketIOService.EXTRA_UN_BLOCK_PARAMTERS, userUnBlocked.toString())
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_UN_BLOCK)
        startService(service)
    }

    private fun checkConnect() {
        val service = Intent(this, SocketIOService::class.java)
        val `object` = JSONObject()
        try {
            `object`.put("my_id", my_id)
            `object`.put("your_id", another_user_id)
            //            socket.emit("check connect", object);
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        service.putExtra(SocketIOService.EXTRA_CHECK_CONNECT_PARAMTERS, `object`.toString())
        service.putExtra(SocketIOService.EXTRA_EVENT_TYPE, SocketIOService.EVENT_TYPE_CHECK_CONNECT)
        startService(service)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        CallProperty.setStatusBarOrScreenStatus(this)
        super.onCreate(savedInstanceState)
        window.setFlags(
            WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN
        )
        binding = DataBindingUtil.setContentView(this, R.layout.activity_user_information)

        LocalBroadcastManager.getInstance(this).registerReceiver(check, IntentFilter(CHEK))
        adapter = MediaAdapter(recyclerDataArrayList, this)
//        EXAMPLE_COUNTER = intPreferencesKey("key")
        dataStore = BaseApp.getInstance().dataStore
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)

        // at last set adapter to recycler view.
        binding.recycleView.layoutManager = layoutManager
        binding.recycleView.adapter = adapter
        initViews()
        initAction()
    }

    @SuppressLint("SetTextI18n")
    private fun initViews() {
        myBase = BaseApp.getInstance()
        classSharedPreferences = myBase.classSharedPreferences
        timeProperties = TimeProperties()
//        blockUserRepo = myBase.blockUserRepo
        val bundle = intent.extras
        userName = bundle!!.getString("name", "Default")
        sn = bundle.getString("special", "Default")
        chatId = bundle.getString("chat_id", "Default")
        another_user_id = bundle.getString("user_id", "Default")
        fcm_token = bundle.getString("fcm_token", "Default")
        imageUrl = bundle.getString("image", "Default")
        blockedFor = bundle.getString("blockedFor", "")
        my_id = classSharedPreferences.user.userId.toString()
//        userInformationViewModelFactory = UserInformationViewModelFactory(blockedFor)
//        userInformationViewModel = ViewModelProvider(this)[UserInformationViewModel::class.java]

//        userInformationViewModel = ViewModelProvider(this,userInformationViewModelFactory)[UserInformationViewModel::class.java]
//        serverApi = ServerApi(this)
        checkConnect()
//        userInformationViewModel.setBlockedFor(blockedFor)
        userInformationViewModel.state.observe(
            this
        ) { s ->
            println("stateee$s")
            if (s != null) {
                if (s == "true") {
                    binding.lastSeen.setText(R.string.connect_now)
                } else if (s == "false") {
                    if (userInformationViewModel.getLastSeen() != "null") {
                        binding.lastSeen.text = resources.getString(R.string.last_seen) + " " + timeProperties.getDateForLastSeen(
                            this,
                            userInformationViewModel.getLastSeen()!!
                                .toLong()
                        )
                    }
                }
            }
        }
        userInformationViewModel.getMedia().observe(
            this
        ) { mediaModelArrayList ->
            if (mediaModelArrayList != null) {
                Log.d("mediaModelArrayList", mediaModelArrayList.size.toString())
                for (media in mediaModelArrayList) {
                    recyclerDataArrayList.add(media)
                }
            }
            adapter!!.notifyDataSetChanged()
        }

        userInformationViewModel.getUserInfo().observe(
            this
        ) { userModel ->
            Log.d("initInfooo", userModel.userName.toString())
            if (userModel != null) {
                binding.txtUserName.text = userModel.userName + " " + userModel.lastName
                if (userModel.phone != null ) {
                    if(userModel.phone!!.isNotEmpty()){

                    val firstString = userModel.phone!!.substring(0, 4)
                    val secondString = userModel.phone!!.substring(4, 7)
                    val thirtyString = userModel.phone!!.substring(7, 10)
                    val lastString = userModel.phone!!.substring(10)
                    binding.txtSpecialNumber.text = "$firstString-$secondString-$thirtyString-$lastString"

//                    if (userModel.image != null) {
//                        Glide.with(circleImageView.context)
//                            .load(AllConstants.imageUrl + userModel.image)
//                            .apply(RequestOptions.placeholderOf(R.drawable.th).error(R.drawable.th))
//                            .into(
//                                circleImageView
//                            )
                    }
                }
            }
        }
        /////////////////check is mute
//        SharedPreferences prefs = getSharedPreferences("muteUsers", MODE_PRIVATE);
//        SharedPreferenceStringLiveData sharedPreferenceStringLiveData = new SharedPreferenceStringLiveData(prefs, "muteUsers", "");
//        sharedPreferenceStringLiveData.getStringLiveData("muteUsers", "").observe(this, cid -> {
////            ArrayList<String> response = new ArrayList<>();
//            Gson gson = new Gson();
//
//
//            String json = prefs.getString("muteUsers", "");
//            Type type = new TypeToken<ArrayList<String>>() {}.getType();
//            if (json != null) {
//                muteList  =  gson.fromJson(json,type );
//
//            }
////            if(response== null){
////                response = new ArrayList<String>();
////            }
//                    for (String s:
//                            muteList) {
//                        if (another_user_id.equals(s)) {
//                            userInformationViewModel.mute.setValue(true);
//                            break;
//                        }
//                    }
//
//        });
        ////////////////
        muteList = classSharedPreferences.muteUsers
        if (muteList != null) {
            for (s in muteList!!) {
                if (another_user_id == s) {
                    userInformationViewModel.setMute(true)
                    break
                }
            }
        }
        userInformationViewModel.loadingMutableLiveData.observe(
            this
        ) { aBoolean ->
            if (aBoolean) {
                binding.imgMute.setImageDrawable(resources.getDrawable(R.drawable.ic_un_mute))
            } else {
                binding.imgMute.setImageDrawable(resources.getDrawable(R.drawable.ic_bell))
            }
        }


//////////////////////////
        userInformationViewModel.isBlocked.observe(
            this
        ) { s ->
            if (s != null) {
                if (s) {
                    sendBlockFor(s)
                    userInformationViewModel.setBlocked(false)
                }
            }
        }
        ////////////
        userInformationViewModel.isUnBlocked().observe(
            this
        ) { s ->
            println("stateee$s")
            if (s != null) {
                if (s) {
                    sendUnBlockFor(s)
                    userInformationViewModel.setUnBlocked(false)
                }
            }
        }
        userInformationViewModel.blockedFor().observe(
            this
        ) { s ->
            val isAnyOneBlock = false
            if (s != null) {
                if (s == my_id || s == "0") {
                    isBlockForMe = true
                   binding.imgAudioCall.isEnabled = false
                    binding.imgVideoCall.isEnabled = false
                } else if (s == another_user_id) {
                    isBlockForMe = false
                    binding.imgAudioCall.isEnabled = false
                    binding.imgVideoCall.isEnabled = false
                } else {
                    isBlockForMe = false
                    binding.imgAudioCall.isEnabled = true
                    binding.imgVideoCall.isEnabled = true
                }
            } else {
                isBlockForMe = false
                binding.imgAudioCall.isEnabled = true
                binding.imgVideoCall.isEnabled = true
            }
        }

        if (imageUrl != null) {
            Glide.with(binding.imageView.context)
                .load(AllConstants.imageUrl + imageUrl)
                .apply(RequestOptions.placeholderOf(R.drawable.th).error(R.drawable.th))
                .into(
                    binding.imageView
                )
        }

        userInformationViewModel.mediaRequest(my_id, another_user_id)
        userInformationViewModel.userInfoRequest( another_user_id)

    }

    private fun initAction() {
        binding.imgMessage.setOnClickListener {
            val bundle = Bundle()
            bundle.putString("reciver_id", another_user_id)
            bundle.putString("sender_id", my_id)
            bundle.putString("fcm_token", fcm_token)

            //        bundle.putString("reciver_id",chatRoomModel.reciverId);
            bundle.putString("name", userName)
            bundle.putString("image", imageUrl)
            bundle.putString("chat_id", chatId)
            bundle.putString("blockedFor", userInformationViewModel.blockedFor().value)
            val intent = Intent(this@UserInformationActivity, ConversationActivity::class.java)
            intent.putExtras(bundle)
            startActivity(intent)
        }
        //////
        binding.imgMore.setOnClickListener { //                block();
            popupMenuExample()
        }
        binding.imgAudioCall.setOnClickListener { //
            if(!checkThereIsOngoingCall()) {

                val intent = Intent(this@UserInformationActivity, RequestCallActivity::class.java)
                intent.putExtra("anthor_user_id", another_user_id)
                intent.putExtra("user_name", userName)
                intent.putExtra("isVideo", false)
                intent.putExtra("fcm_token", fcm_token)
                intent.putExtra("image_profile", imageUrl)
                startActivity(intent)
            }
        }
        binding.imgVideoCall.setOnClickListener {
            if(!checkThereIsOngoingCall()) {

                val intent = Intent(this@UserInformationActivity, RequestCallActivity::class.java)
                //                Intent intent = new Intent(ConversationActivity.this, CompleteActivity.class);
                intent.putExtra("anthor_user_id", another_user_id)
                intent.putExtra("user_name", userName)
                intent.putExtra("isVideo", true)
                intent.putExtra("fcm_token", fcm_token)
                intent.putExtra("image_profile", imageUrl)
                startActivity(intent)
            }
        }
        binding.imgMute.setOnClickListener { //                System.out.println("muteeeeeeeeeeeee");
            //                JSONObject jsonObject = new JSONObject();
            if (userInformationViewModel.loadingMutableLiveData.value!!) {
                muteList!!.remove(another_user_id)
                userInformationViewModel.setMute(false)
            } else {
                muteList!!.add(another_user_id)
                userInformationViewModel.setMute(true)
            }
            classSharedPreferences.muteUsers = muteList
        }

        binding.imageView.setOnClickListener {
                val dialog = Dialog(this)
                dialog.setContentView(R.layout.dialog_image_cht)
                dialog.setTitle("Title...")
                dialog.window!!
                    .setLayout(
                        ViewGroup.LayoutParams.FILL_PARENT,
                        ViewGroup.LayoutParams.FILL_PARENT
                    )
                val image: PhotoView = dialog.findViewById(R.id.photo_view)
                Glide.with(image.context).load(AllConstants.imageUrl + imageUrl).centerCrop()
                    .into(image)
                dialog.show()
            }

    }

    private fun popupMenuExample() {
        val p = PopupMenu(this, binding.linerMore)
        p.menuInflater.inflate(R.menu.main_menu, p.menu)
        if (isBlockForMe) {
            p.menu.findItem(R.id.block).isVisible = false
            p.menu.findItem(R.id.unBlock).isVisible = true
        } else {
            p.menu.findItem(R.id.block).isVisible = true
            p.menu.findItem(R.id.unBlock).isVisible = false
        }
        p.setOnMenuItemClickListener { item ->
            val userModel = UserModel(another_user_id, userName, userName, "", "", sn, imageUrl, "")
            when (item.itemId) {
                R.id.block -> {
                    val dialog = AlertDialog.Builder(this@UserInformationActivity)
                    dialog.setTitle(R.string.alert_block_user)
                    dialog.setPositiveButton(
                        R.string.block
                    ) { dialog, which -> //                                        serverApi.block(my_id,userModel);
                        userInformationViewModel.sendBlockRequest(
                            my_id,
                            another_user_id
                        )
                    }
                    dialog.setNegativeButton(
                        R.string.cancel
                    ) { dialog, which -> dialog.dismiss() }
                    val alertDialog = dialog.create()
                    alertDialog.show()
                }
                R.id.unBlock -> {
                    val dialogUnBlock =
                        AlertDialog.Builder(this)
                    dialogUnBlock.setTitle(R.string.alert_unblock_user)
                    dialogUnBlock.setPositiveButton(
                        R.string.Unblock
                    ) { dialog, which -> //                                    serverApi.unbBlockUser(my_id,userModel);
                        userInformationViewModel.sendUnBlockRequest(
                            my_id,
                            another_user_id
                        )
                    }
                    dialogUnBlock.setNegativeButton(
                        R.string.cancel
                    ) { dialog, which -> dialog.dismiss() }
                    val alertUnBlockDialog = dialogUnBlock.create()
                    alertUnBlockDialog.show()
                }
            }
            true
        }
        p.show()
    }

    override fun onDestroy() {
        super.onDestroy()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(check)
    }

    companion object {
        const val CHEK = "ConversationActivity.CHECK_CONNECT"
    }


}